src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivityTagsCollection.cs

```diff
@@ -0,0 +1,261 @@
+        ///     - If the key previousely existed in the collection and the value is not null, the value will replace the old value stored in the collection.
+        ///     - Otherwise, a new item will get added to the collection.
+        /// </summary>
+        /// <value>Object mapped to the key</value>
+        public object this[string key]
+        {
+            get
+            {
+                int index = _list.FindIndex(kvp => kvp.Key == key);
+                return index < 0 ? null! : _list[index].Value;
+            }
+
+            set
+            {
+                if (key == null)
+                {
+                    throw new ArgumentNullException(nameof(key));
+                }
+
+                int index = _list.FindIndex(kvp => kvp.Key == key);
+                if (value == null)
+                {
+                    if (index >= 0)
+                    {
+                        _list.RemoveAt(index);
+                    }
+                    return;
+                }
+
+                if (index >= 0)
+                {
+                    _list[index] = new KeyValuePair<string, object>(key, value);
+                }
+                else
+                {
+                    _list.Add(new KeyValuePair<string, object>(key, value));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get the list of the keys of all stored tags.
+        /// </summary>
+        public ICollection<string> Keys
+        {
+            get
+            {
+                List<string> list = new List<string>(_list.Count);
+                foreach (KeyValuePair<string, object> kvp in _list)
+                {
+                    list.Add(kvp.Key);
+                }
+                return list;
+            }
+        }
+
+        /// <summary>
+        /// Get the list of the values of all stored tags.
+        /// </summary>
+        public ICollection<object> Values
+        {
+            get
+            {
+                List<object> list = new List<object>(_list.Count);
+                foreach (KeyValuePair<string, object> kvp in _list)
+                {
+                    list.Add(kvp.Value);
+                }
+                return list;
+            }
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether the collection is read-only.
+        /// </summary>
+        public bool IsReadOnly => false;
+
+        /// <summary>
+        /// Gets the number of elements contained in the collection.
+        /// </summary>
+        public int Count => _list.Count;
+
+        /// <summary>
+        /// Adds an tag with the provided key and value to the collection.
+        /// This collection doesn't allow adding two tags with the same key.
+        /// </summary>
+        /// <param name="key">The tag key.</param>
+        /// <param name="value">The tag value.</param>
+        public void Add(string key, object value)
```

Comment:
As the behavior is that of a `Set` method than a `Add` method, shouldn't we call this as Set?

Reply:
No, Add behavior is different than Set behavior. Add will always add a new entry even if there is previous entry with the same key is added. Also add can allow null value. This is the whole point we are adding 2 APIs and not just one.